File: src/main/java/com/enablehr/challenge/java/controller/TodoController.java
package com.enablehr.challenge.java.controller;

import com.enablehr.challenge.java.entity.Todo;
import com.enablehr.challenge.java.entity.TodoResp;
import com.enablehr.challenge.java.repository.TodoRepository;
import com.enablehr.challenge.java.service.TodoService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.annotation.PostConstruct;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;


@RestController
@CrossOrigin
@RequestMapping("/todo")
public class TodoController {

  @Autowired
  TodoRepository todoRepository;

  @Autowired
  TodoService todoService;

  @PostMapping("/add")
  public TodoResp create(@RequestBody Todo body) {
    return todoService.create(body.getText(), body.getParentId());
  }

  @GetMapping("/list")
  public List<TodoResp> list() {
    return todoService.list();
  }

  @PutMapping("/update")
  public TodoResp update(@RequestBody Todo body) {
    return todoService.update(body);
  }

  @DeleteMapping("/delete")
  public void delete(@RequestBody Todo body) {
    todoService.delete(body.getId());
  }

  @PostMapping("/complete")
  public TodoResp markAsComplete(@RequestBody Todo body) {
    return todoService.markAsComplete(body.getId());
  }

  @PostMapping("/batch/clear")
  public List<TodoResp> clearCompleted() {
    return todoService.clearCompleted();
  }

  @PostMapping("/batch/complete")
  public List<TodoResp> completeAll() {
    return todoService.completeAll();
  }

  @PostConstruct
  public void init() {

    Stream.of("Use Redux", "Use React", "Pass the Test")
          .forEach(text -> {
            final Todo todo = new Todo();
            todo.setCompleted(Boolean.FALSE);
            todo.setText(text);
            todo.setParentId(0);
            todoRepository.save(todo);
          });
  }
}

File: src/main/kotlin/com/enablehr/challenge/kotlin/controller/TodoController.kt
package com.enablehr.challenge.kotlin.controller

import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController
import javax.annotation.PostConstruct

@RestController
@CrossOrigin
class TodoController {

    internal companion object {
        const val COMPLETED = "completed"
        const val ID = "id"
        const val TEXT = "text"
    }

    internal val todos: MutableList<MutableMap<String, Any?>> = mutableListOf()

    private var counter = 0

    @RequestMapping(path = arrayOf("/todos/clear"), method = arrayOf(RequestMethod.POST))
    fun clearCompleted(): List<Map<String, Any?>> =
        todos.apply { removeIf { todo -> todo[COMPLETED] == true } }

    @RequestMapping(path = arrayOf("/todos/complete"), method = arrayOf(RequestMethod.POST))
    fun completeAll(): List<Map<String, Any?>> {

        val complete = todos.map { todo -> todo[COMPLETED] }
            .any { completed -> completed == false }

        return todos.apply {
            forEach { todo -> todo.put(COMPLETED, complete) }
        }
    }

    @RequestMapping(path = arrayOf("/todo"), method = arrayOf(RequestMethod.POST))
    fun create(@RequestBody body: Map<String, Any?>): Map<String, Any?> =
        mutableMapOf(ID to counter++,
                     COMPLETED to false,
                     TEXT to body[TEXT])
            .also { todos.add(it) }

    @RequestMapping(path = arrayOf("/todo"), method = arrayOf(RequestMethod.DELETE))
    fun delete(@RequestBody body: Map<String, Any?>) {
        body[ID]?.let { id -> todos.removeIf { id == it[ID] } }
    }

    @PostConstruct
    fun init() {
        todos.clear()
        counter = 0
        listOf("Use Redux", "Use React", "Pass the Test")
            .map { mutableMapOf<String, Any?>(ID to counter++,
                                              COMPLETED to false,
                                              TEXT to it) }
            .forEach { todos.add(it) }
    }

    @RequestMapping(path = arrayOf("/todos"), method = arrayOf(RequestMethod.GET))
    fun list(): List<Map<String, Any?>> =
        todos

    @RequestMapping(path = arrayOf("/todo/complete"), method = arrayOf(RequestMethod.POST))
    fun markAsComplete(@RequestBody body: Map<String, Any?>): Map<String, Any?>? =
        findById(body[ID] as Int?)
            ?.also {
                val completed = it[COMPLETED] as Boolean
                it[COMPLETED] = !completed
            }

    @RequestMapping(path = arrayOf("/todo"), method = arrayOf(RequestMethod.PUT))
    fun update(@RequestBody body: Map<String, Any?>): Map<String, Any?>? =
        findById(body[ID] as Int?)
            ?.also { it[TEXT] = body[TEXT] }

    private fun findById(id: Int?): MutableMap<String, Any?>? =
        when (id) {
            null -> null
            else -> todos.find { id == it[ID] }
        }

}
